File: core/src/main/old/core/Feature.java
package com.gitlab.aecsocket.sokol.core;

import com.gitlab.aecsocket.minecommons.core.translation.Localizer;
import com.gitlab.aecsocket.sokol.core.registry.Keyed;
import net.kyori.adventure.text.Component;
import org.spongepowered.configurate.ConfigurationNode;
import org.spongepowered.configurate.serialize.SerializationException;

import java.util.List;
import java.util.Locale;
import java.util.Optional;

public interface Feature<
        F extends FeatureInstance<D, ?>,
        D
> extends Keyed {
    int PRIORITY_DEFAULT = 0;
    int PRIORITY_LOW = -1000;
    int PRIORITY_HIGH = 1000;

    F load(D data);

    void configure(ConfigurationNode config) throws SerializationException;

    static String renderKey(String id) { return "feature." + id + ".name"; }
    static String renderDescriptionKey(String id) { return "feature." + id + ".description"; }

    @Override
    default Component render(Locale locale, Localizer lc) {
        return lc.safe(locale, renderKey(id()));
    }

    default Component renderDescription(Locale locale, Localizer lc) {
        return lc.safe(locale, renderDescriptionKey(id()));
    }

    Optional<List<Component>> renderConfig(Locale locale, Localizer lc);
}

File: core/src/main/kotlin/com/github/aecsocket/sokol/core/Feature.kt
package com.github.aecsocket.sokol.core

import com.github.aecsocket.alexandria.core.keyed.Keyed
import com.github.aecsocket.glossa.core.I18N
import com.github.aecsocket.glossa.core.Localizable
import com.github.aecsocket.sokol.core.event.NodeEvent
import com.github.aecsocket.sokol.core.nbt.CompoundBinaryTag
import com.github.aecsocket.sokol.core.nbt.TagSerializable
import com.github.aecsocket.sokol.core.rule.Rule
import com.github.aecsocket.sokol.core.stat.ApplicableStats
import com.github.aecsocket.sokol.core.stat.Stat
import net.kyori.adventure.key.Key
import net.kyori.adventure.text.Component
import org.spongepowered.configurate.ConfigurationNode

interface Feature<
    P : Feature.Profile<*>
> : Keyed, Localizable<Component> {
    val statTypes: Map<Key, Stat<*>>

    val ruleTypes: Map<Key, Class<Rule>>

    fun createProfile(node: ConfigurationNode): P

    override fun localize(i18n: I18N<Component>) =
        i18n.safe("feature.$id")

    interface Profile<D : Data<*>> : Keyed {
        val type: Feature<*>
        override val id get() = type.id

        fun createData(): D

        fun createData(node: ConfigurationNode): D

        fun createData(tag: CompoundBinaryTag): D
    }

    interface Data<S : State<S, *, *>> : TagSerializable, Keyed {
        val type: Feature<*>
        val profile: Profile<*>
        override val id get() = type.id

        fun createState(): S

        fun serialize(node: ConfigurationNode)
    }

    interface State<
        S : State<S, D, C>,
        D : Data<S>,
        C : FeatureContext<*, *, *>
    > : TagSerializable, Keyed {
        val type: Feature<*>
        val profile: Profile<*>
        override val id get() = type.id

        fun asData(): D

        fun resolveDependencies(get: (String) -> S?) {}

        fun createStats(): List<ApplicableStats> = emptyList()

        fun onEvent(event: NodeEvent, ctx: C)
    }
}

interface FeatureContext<
    S : TreeState,
    H : NodeHost,
    N
> where N : DataNode, N : Node.Mutable<N> {
    val state: S
    val host: H
    val node: N

    fun writeNode(action: N.() -> Unit)
}
